File: helloworld.vaadin/src/main/scala/ch/codecraft/helloworld/vaadin/MyVaadinApplication.scala
package ch.codecraft.helloworld.vaadin

import com.vaadin.Application
import com.vaadin.ui._

class MyVaadinApplication extends Application {

  def init(): Unit = {
     setMainWindow(new Window("My Vaadin Application"))
     val button = new Button("Click Me")
     
     //TODO find out how scala.swing handles SAM classes
     button.addListener(new Button.ClickListener() {
       def buttonClick(event: Button#ClickEvent): Unit = {
         getMainWindow.addComponent(new Label("Thank you for clicking"))
       }
     })
     //FIXME button.addListener(_ => getMainWindow.addComponent(new Label("Thank you for clicking")))

     getMainWindow.addComponent(button)
  }

}

File: helloworld.vaadin/src/main/java/ch/codecraft/helloworld/vaadin/MyVaadinApplication.java
package ch.codecraft.helloworld.vaadin;

import java.util.Date;

import com.vaadin.Application;
import com.vaadin.ui.Button;
import com.vaadin.ui.Button.ClickEvent;
import com.vaadin.ui.Label;
import com.vaadin.ui.Window;

public class MyVaadinApplication extends Application {

    private static final long serialVersionUID = 1L;

    @Override
    public void init() {
        final Window main = new Window("Hello window");
        setMainWindow(main);
        main.addComponent(new Label("Hello World!"));

        main.addComponent(new Button("what time is it?", new Button.ClickListener() {
            private static final long serialVersionUID = 1L;

            @Override
            public void buttonClick(ClickEvent event) {
                main.showNotification("Time now " + new Date());
            }
        }));
    }

}
